### ISSUE TITLE
Allow {} inside of an argument for ShellCommand configtype (e.g. editor setting)

### ISSUE THREAD

**Issue by Earnestly** [The-Compiler/qutebrowser#1092]
The `:set general editor` improperly relies on inconsistent semantics of terminals such as `xterm -e` and `urxvt -e` to make assumptions about which argument goes to the command and which goes to the terminal emulator.  They add quriks which allows the use of `xterm -e vim file` to send the `file` to `vim` instead of `xterm`.

This is inconsistent as every other command-line interface will be subject to wordsplitting requiring quoting to avoid it.  An example is livestreamer
```
% livestreamer --player mpv --player-args "--title='spaces in title' {filename}" http://stream.url
```

Obviously in the above, `--title` and `{filename}` should *not* be sent to `livestreamer` itself but passed to the `mpv` command.  `xterm` and `urxvt` both subvert this consistency in the name of convenience.

Other terminals (all libvte based terminals) do not repeat this flaw such as termite (https://github.com/thestinger/termite/issues/246) and as a result using `:set general editor "termite -e vim {}"` will pass the expansion of `{}` to `termite` instead of `vim`.

**Comment by The-Compiler** [comment]
As discussed in IRC, the quoting isn't the issue, e.g. setting the value to `urxvt -e vim '/tmp/foo bar' {}` works just fine.

The issue is that `{}` isn't allowed inside an argument (after qutebrowser's shell-like splitting of the setting value), but only as an argument of its own.

That behavior was actually introduced explicitly in 156d63b4b6eb25d852bf970d28f4628c9c1d206c - I don't remember why though, and I didn't mention that in the commit message...

Also I think there are different places handling `{}` replacement - probably some function in `utils.py` should be added instead.

**Comment by Earnestly** [comment]
The issue is *how* the arguments are passed to the software, most software does not do special positional handling of command-line arguments.  It is the difference between: `execl("/usr/bin/termite", "-e", "vim", "file", 0);` and `execl("/usr/bin/termite", "-e", "vim file", 0);`

So of course `urxvt` will work, it has been *designed* to handle `-e` in this manner!

This is behaviour was not introduced by that commit, the behaviour has been part of the codebase since the functionality was added.

**Comment by Earnestly** [comment]
A syntactic mechanism (or otherwise) for allowing users to group parts of a command line into a single entity is required, this is typically done using quoting.  

The result is that `cmd "foo bar"` would consistent of the array `["cmd, "foo bar"]` and not `["cmd", "foo", "bar"]`.

There will be issues with escaping which is usually handled by either doubling up (i.e. most grammar syntaxes) or C-like escaping with `\`.

**Comment by Earnestly** [comment]
Commands like `:spawn` behave correctly here for example because as the note in the documentation says, everything is taken literally.  Although I'm not sure about the mechanism used to interpret this literal text, a shell?

**Comment by The-Compiler** [comment]
Again - you *can* already quote arguments with spaces, as I demonstrated above with the `/tmp/foo bar` example. The issue is that having `{}` as single argument is enforced, which shouldn't be.

This is done via Python's [shlex.split](https://docs.python.org/3.5/library/shlex.html).

**Comment by Earnestly** [comment]
Perhaps you should use something other than `urxvt` because your example that is supposed to work, does not.

**Comment by The-Compiler** [comment]
I also don't know how else to explain this. You can't test this with `termite` because you *need* to have a `{}` argument, which would get passed to `termite`, and I presume it doesn't like that.

Here is a proof of concept fix, which makes external editing work with `termite` with `general -> editor` set to `termite -e "vim {}" ` (so, `:set general editor 'termite -e "vim {}" '`):

```diff
diff --git a/qutebrowser/config/configtypes.py b/qutebrowser/config/configtypes.py
index 4066c4c..f8c442a 100644
--- a/qutebrowser/config/configtypes.py
+++ b/qutebrowser/config/configtypes.py
@@ -1076,9 +1076,9 @@ class ShellCommand(BaseType):
             shlex.split(value)
         except ValueError as e:
             raise configexc.ValidationError(value, str(e))
-        if self.placeholder and '{}' not in self.transform(value):
-            raise configexc.ValidationError(value, "needs to contain a "
-                                            "{}-placeholder.")
+        #if self.placeholder and '{}' not in self.transform(value):
+        #    raise configexc.ValidationError(value, "needs to contain a "
+        #                                    "{}-placeholder.")
 
     def transform(self, value):
         if not value:
diff --git a/qutebrowser/misc/editor.py b/qutebrowser/misc/editor.py
index 4211fe1..e367591 100644
--- a/qutebrowser/misc/editor.py
+++ b/qutebrowser/misc/editor.py
@@ -124,6 +124,7 @@ class ExternalEditor(QObject):
         self._proc.error.connect(self.on_proc_error)
         editor = config.get('general', 'editor')
         executable = editor[0]
-        args = [self._filename if arg == '{}' else arg for arg in editor[1:]]
+        #args = [self._filename if arg == '{}' else arg for arg in editor[1:]]
+        args = [arg.replace('{}', self._filename) for arg in editor[1:]]
         log.procs.debug("Calling \"{}\" with args {}".format(executable, args))
         self._proc.start(executable, args)
```

As you can see, no quoting fixes involved. Quoting works fine already.

**Comment by The-Compiler** [comment]
Fixed via #1263.
